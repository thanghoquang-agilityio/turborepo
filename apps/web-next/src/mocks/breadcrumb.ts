export const MOCK_BREADCRUMB_ITEMS = [
  { name: 'Products', link: 'products' },
  { name: 'Detail Product', link: '7' },
]
