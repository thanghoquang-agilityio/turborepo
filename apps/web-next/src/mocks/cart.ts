import { CartItemResponse, LegacyCartItemResponse } from '@/types'
import { MOCK_PRODUCT_VARIANTS } from './product'

export const MOCK_CART_ITEM = {
  id: '1',
  variantId: '1',
  name: 'Embrace Sideboard',
  color: 'wheat',
  size: 'small',
  quantityDefault: 2,
  stock: 4,
  price: 53,
  star: 1,
  comments: 1,
  tax: 1,
  discount: 1,
}

// New flattened cart response structure
export const MOCK_CART_RESPONSE: CartItemResponse[] = [
  {
    id: 1,
    documentId: 'cart-item-1',
    size: 'small',
    quantity: 2,
    createdAt: '2024-01-01T00:00:00.000Z',
    updatedAt: '2024-01-01T00:00:00.000Z',
    publishedAt: '2024-01-01T00:00:00.000Z',
    productVariantId: MOCK_PRODUCT_VARIANTS[0],
    userId: 1,
  },
  {
    id: 2,
    documentId: 'cart-item-2',
    size: 'large',
    quantity: 2,
    createdAt: '2024-01-01T00:00:00.000Z',
    updatedAt: '2024-01-01T00:00:00.000Z',
    publishedAt: '2024-01-01T00:00:00.000Z',
    productVariantId: MOCK_PRODUCT_VARIANTS[1],
    userId: 1,
  },
]

// Legacy cart response for backward compatibility
export const MOCK_CART_RESPONSE_LEGACY: LegacyCartItemResponse[] = [
  {
    id: '1',
    attributes: {
      productVariantId: {
        data: {
          id: '1',
          attributes: {
            size: 'small,medium,large,extra large,XXL',
            color: 'slate blue',
            stock: '2,10,0,23,20',
            tax: 12,
          },
        },
      },
      size: 'small',
      quantity: 2,
    },
  },
  {
    id: '2',
    attributes: {
      productVariantId: {
        data: {
          id: '1',
          attributes: {
            size: 'small,medium,large,extra large,XXL',
            color: 'slate blue',
            stock: '20,10,20,23,20',
            tax: 12,
          },
        },
      },
      size: 'large',
      quantity: 2,
    },
  },
]

export const MOCK_ORDER_SUMMARY = {
  subTotal: 396,
  itemTotal: 3,
}
