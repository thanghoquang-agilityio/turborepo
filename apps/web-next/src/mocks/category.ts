export const MOCK_CATEGORY_LIST = [
  {
    name: 'Kid',
    value: 'kid',
    total: 18,
  },
  {
    name: 'Man',
    value: 'man',
    total: 12,
  },
  {
    name: 'Woman',
    value: 'woman',
    total: 23,
  },
  {
    name: 'Sport',
    value: 'sport',
    total: 34,
  },
]

export const MOCK_CATEGORY_RESPONSE = [
  {
    name: 'Kid',
    value: 'kid',
    products: {
      data: [],
    },
  },
  {
    name: 'Man',
    value: 'man',
    products: {
      data: [],
    },
  },
  {
    name: 'Woman',
    value: 'woman',
    products: {
      data: [],
    },
  },
  {
    name: 'Sport',
    value: 'sport',
    products: {
      data: [],
    },
  },
]
