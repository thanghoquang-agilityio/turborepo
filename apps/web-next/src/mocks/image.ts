import { SRC_IMAGE_NOT_AVAILABLE } from '@/constants'

export const MOCK_IMAGE = {
  data: {
    id: '24',
    attributes: {
      name: 'default',
      url: SRC_IMAGE_NOT_AVAILABLE,
    },
  },
}

export const MOCK_LIST_IMAGE_RESPONSE = [
  {
    name: 'default1',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
  {
    name: 'default2',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
  {
    name: 'default3',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
  {
    name: 'default4',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
]
