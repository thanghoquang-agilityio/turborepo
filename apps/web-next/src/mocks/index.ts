export * from './breadcrumb'
export * from './cart'
export * from './category'
export * from './image'
export * from './product'
export * from './radio-group'
export * from './user'
