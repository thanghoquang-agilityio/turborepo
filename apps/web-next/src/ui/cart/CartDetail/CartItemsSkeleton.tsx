import { Skeleton } from '@nextui-org/react'

import { PAGE_SIZE_DEFAULT, borderStyle } from '@/constants'
// Styles
import '@/styles/card-list.css'
import '@/styles/card.css'

export const CartItemsSkeleton = () => {
  return (
    <div
      className={`card-list mx-auto grid w-fit grid-cols-2 gap-x-[20px] gap-y-[30px] 2xl:grid-cols-3`}
    >
      {Array(PAGE_SIZE_DEFAULT)
        .fill(0)
        .map((_, index) => (
          <div
            key={`card-product-skeleton-${index}`}
            className={`card flex max-w-[305px] flex-col items-center overflow-hidden rounded-large border-medium border-background-300 pb-5 pt-3`}
          >
            <Skeleton className='bg-default-200 h-[26px] w-[170px] rounded-large' />
            <Skeleton className='bg-default-200 mt-1 h-[18px] w-[100px] rounded-large' />
            <Skeleton className='card-image bg-default-200 mb-5 mt-4 h-[185px] w-full md:h-[303px] md:w-[305px]' />
            <Skeleton className='bg-default-200 h-[38px] w-[150px] rounded-large' />

            <div
              className={`relative mt-10 flex w-full justify-between px-4 ${borderStyle} before:top-[-20px]`}
            >
              <Skeleton className='bg-default-200 h-[26px] w-[90px] rounded-large' />
              <Skeleton className='bg-default-200 h-[26px] w-[100px] rounded-large' />
            </div>
          </div>
        ))}
    </div>
  )
}
