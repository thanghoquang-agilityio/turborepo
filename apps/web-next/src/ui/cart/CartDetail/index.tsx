'use client'

import { Suspense } from 'react'

import dynamic from 'next/dynamic'

import { useOrderSummary } from '@/hooks'
// Styles
import '@/styles/card-list.css'
// Types
import { CartItemResponse } from '@/types'

// Components
const CartItem = dynamic(() => import('../CartItem'))
const OrderSummary = dynamic(() => import('../OrderSummary'))

interface CartDetailProps {
  cartItems: CartItemResponse[]
}

const CartDetail = ({ cartItems }: CartDetailProps) => {
  const { totalAmount, totalItems, addOptimisticCartItems } =
    useOrderSummary(cartItems)

  return (
    <div className='container flex w-full max-w-full flex-col items-center gap-[40px] py-8 lg:flex-row lg:items-start lg:py-12 xl:py-16'>
      <div
        className={`card-list mx-auto grid w-fit grid-cols-2 gap-x-[20px] gap-y-[30px] 2xl:grid-cols-3`}
      >
        {cartItems.map((cartItem) => {
          const { 
            id = '', 
            quantity = 0, 
            size = '', 
            productVariantId 
          } = cartItem
          
          const {
            id: variantId = '',
            color = '',
            size: sizeList = '',
            stock: stockList = '',
            tax = 0,
          } = productVariantId || {}

          // Handle product relationship in new API structure
          // These could come from:
          // 1. Direct properties on productVariantId (if API includes product data)
          // 2. Separate API calls to get product details
          // 3. Extended cart response with product information
          const name = '' // productVariantId?.product?.name || cartItem.productName || ''
          const image = '' // productVariantId?.product?.image?.url || cartItem.productImage || ''
          const discount = 0 // productVariantId?.product?.discount || cartItem.productDiscount || 0
          const price = 0 // productVariantId?.product?.price || cartItem.productPrice || 0

          const sizes = sizeList.split(',')
          const stocks = stockList.split(',')
          const sizeIndex = sizes.indexOf(size)
          const stock = Number(stocks[sizeIndex])

          return (
            <Suspense key={`cart-item-suspense-${id}`}>
              <CartItem
                key={`cart-item${id}`}
                variantId={variantId.toString()}
                id={id.toString()}
                quantityDefault={quantity}
                size={size}
                color={color}
                name={name}
                image={image}
                price={price}
                tax={tax}
                discount={discount}
                stock={stock}
                addOptimisticCartItems={addOptimisticCartItems}
              />
            </Suspense>
          )
        })}
      </div>
      <div className='w-[305px]'>
        <Suspense>
          <OrderSummary subTotal={totalAmount} itemTotal={totalItems} />
        </Suspense>
      </div>
    </div>
  )
}

export default CartDetail
