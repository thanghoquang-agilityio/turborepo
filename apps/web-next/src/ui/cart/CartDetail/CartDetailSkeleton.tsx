import { Skeleton } from '@repo/next-ui'

// Styles
import '@/styles/card-list.css'
import '@/styles/card.css'

import { CartItemsSkeleton } from './CartItemsSkeleton'

export const CartDetailSkeleton = () => (
  <div className='container flex w-full max-w-full flex-col items-center gap-[40px] py-8 lg:flex-row lg:items-start lg:py-12 xl:py-16'>
    <CartItemsSkeleton />
    <Skeleton className='bg-default-200 mb-8 h-[464px] w-[304px] min-w-[304px] rounded-medium border-medium border-background-300 p-4' />
  </div>
)
