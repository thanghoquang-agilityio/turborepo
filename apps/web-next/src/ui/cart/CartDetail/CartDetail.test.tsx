import { render } from '@testing-library/react'

import { MOCK_CART_RESPONSE } from '@/mocks'
import type { CartItemResponse } from '@/types'

import CartDetail from '.'

// Mock the useOptimistic hook from React
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useOptimistic: jest.fn((initialState, updateFn) => {
    let state = initialState

    const updateState = (newItem: CartItemResponse) => {
      state = updateFn(state, newItem)
    }

    return [state, updateState]
  }),
}))

describe('CartDetail', () => {
  it('should match snapshot', () => {
    const { container } = render(<CartDetail cartItems={MOCK_CART_RESPONSE} />)
    expect(container).toMatchSnapshot()
  })
})
