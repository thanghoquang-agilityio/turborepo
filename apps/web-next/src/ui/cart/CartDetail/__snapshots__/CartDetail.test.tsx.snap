// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDetail should match snapshot 1`] = `
<div>
  <div
    class="container flex lg:flex-row flex-col gap-[40px] lg:items-start items-center w-full max-w-full xl:py-16 lg:py-12 py-8"
  >
    <div
      class="card-list grid grid-cols-2 gap-x-[20px] gap-y-[30px] w-fit mx-auto 2xl:grid-cols-3"
    />
    <div
      class="w-[305px]"
    />
  </div>
</div>
`;
